.github/workflows/release.yml

```diff
@@ -50,72 +50,161 @@ on:
+        RELEASE_VERSION: ${{ steps.release_version.outputs.RELEASE_VERSION }}
     steps:
+      - name: Export release version
+        id: release_version
+        # transform tag format `v<version` to `<version>`
+        run: |
+          RELEASE_VERSION="${RELEASE_TAG_VERSION:1}"
+          echo "RELEASE_VERSION=${RELEASE_VERSION}" >> "$GITHUB_OUTPUT"
+
       - uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab  # v3.5.2
         with:
-          fetch-depth: 0
+          ref: ${{ env.RELEASE_COMMIT }}
+
+      # We use a pinned version of Poetry to be certain it won't modify source code before we create a hash
       - name: Install poetry
-        run: pipx install poetry
+        run: pipx install git+https://github.com/python-poetry/poetry@68b88e5390720a3dd84f02940ec5200bfce39ac6 # v1.5.0
+
+      - name: Bump package version
+        id: versioning
+        run: poetry version "${RELEASE_VERSION}"
+        env:
+          RELEASE_VERSION: ${{ steps.release_version.outputs.RELEASE_VERSION}}
+
+      - name: Create integrity hash
+        id: integrity
+        run: echo "INTEGRITY_HASH=${SOURCE_CODE_HASH}" >> "$GITHUB_OUTPUT"
+        env:
+          # paths to hash and why they're important to protect
+          #
+          # aws_lambda_powertools/  - source code
+          # pyproject.toml          - project metadata
+          # poetry.lock             - project dependencies
+          # layer/                  - layer infrastructure and pipeline
+          # .github/                - github scripts and actions used in the release
+          # docs/                   - user guide documentation
+          # examples/               - user guide code snippets
+          SOURCE_CODE_HASH: ${{ hashFiles('aws_lambda_powertools/**', 'pyproject.toml', 'poetry.lock', 'layer/**', '.github/**', 'docs/**', 'examples/**')}}
+
+      - name: Upload sealed source code
+        uses: ./.github/actions/upload-artifact
+        with:
+          name: ${{ steps.integrity.outputs.INTEGRITY_HASH }}
+          path: .
+
+
+  # This job runs our automated test suite, complexity and security baselines
+  # it ensures previously merged have been tested as part of the pull request process
+  #
+  # NOTE
+  #
+  # we don't upload the artifact after tested to prevent no dependencies tamper our source code
+  quality_check:
+    needs: seal
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+    steps:
+      # NOTE: we need actions/checkout to configure git first (pre-commit hooks in make dev)
+      - uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab  # v3.5.2
+        with:
+          ref: ${{ env.RELEASE_COMMIT }}
+
+      - name: Restore sealed source code
+        uses: ./.github/actions/download-artifact
+        with:
+          name: ${{ needs.seal.outputs.INTEGRITY_HASH }}
+          path: .
+
+      - name: Debug cache restore
+        run: cat pyproject.toml
```

Comment:
I left this one thinking it'd be useful in case something ever goes wrong in any other stage, we'd know what's the bumped version and project metadata looked like in the pipeline (no secret leak here).

Happy to remove if you think otherwise. My gut tells me to leave as debugging GH Actions is always a pain after it fails.